cpp_utils/include/cpp_utils/types/IniciableSingleton.hpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file Singleton.hpp
+ *
+ * This file contains class Singleton definition.
+ */
+
+#pragma once
+
+#include <memory>
+#include <shared_mutex>
+
+namespace eprosima {
+namespace utils {
+
+/**
+ * @brief This auxiliary class allows to create a Singleton class that can be initialize statically (before main).
+ *
+ * In order to create a Singleton of a class T that must be initialized before use, this class helps to define and
+ * implement the class T as a normal class and then use it as a Singleton by using the type IniciableSingleton<T>.
+ *
+ * @note for more information about the Singleton class, refer to \c Singleton .
+ *
+ * @tparam T type of the value that will be converted to Singleton.
+ * @tparam Index identifier of a specific Singleton element.
+ *
+ * @example
+ *   class Object;  // A class that represents a generic object, but has no default constructor.
+ *   using InitializedObject = IniciableSingleton<Object>;
+ *   auto __rubbish_ = InitializedObject::initialize<Object ctor arguments>( { initialization arguments } );
```

Comment:
```suggestion
 *   auto __rubbish_ = InitializedObject::initialize<Object constructor arguments>( { initialization arguments } );
```

Reply:
okey this time, but ctor and dtor have come to stay. Even ISO uses it: https://isocpp.org/wiki/faq/ctors#overview-ctors